sklearn/manifold/tests/test_t_sne.py

```diff
@@ -262,3 +422,128 @@ def test_chebyshev_metric():
+    """Raise a warning if fit and transform encountered the same data."""
+    with warnings.catch_warnings(record=True) as w:
+        warnings.simplefilter("always")
+        random_state = check_random_state(0)
+        X = random_state.randn(100, 2)
+        tsne = TSNE(n_components=2, perplexity=2, learning_rate=100.0,
+                    random_state=0, method='barnes_hut')
+        tsne.fit(X)
+        tsne.transform(X)
+        m = str(w[0].message)
+    assert "The transform input appears to be similar" in m
+
+
+def test_barnes_hut_angle():
+    """When Barnes-Hut's angle=0 this corresponds to the exact method."""
+    angle = 0.0
+    perplexity = 10
+    n_samples = 100
+    n_components = 2
+    n_features = 5
+    degrees_of_freedom = float(n_components - 1.0)
+
+    random_state = check_random_state(0)
+    distances = random_state.randn(n_samples, n_features).astype(np.float32)
+    distances = distances.dot(distances.T)
+    np.fill_diagonal(distances, 0.0)
+    params = random_state.randn(n_samples, n_components)
+    P = _joint_probabilities(distances, perplexity, False)
+    kl, gradex = _kl_divergence(params, P, degrees_of_freedom, n_samples,
+                                n_components)
+
+    k = n_samples - 1
+    bt = BallTree(distances)
+    distances_nn, neighbors_nn = bt.query(distances, k=k+1)
+    neighbors_nn = neighbors_nn[:, 1:]
+    Pbh = _joint_probabilities_nn(distances, neighbors_nn,
+                                  perplexity, False)
+    kl, gradbh = _kl_divergence_bh(params, Pbh, neighbors_nn,
+                                   degrees_of_freedom, n_samples, n_components,
+                                   angle=angle, skip_num_points=0,
+                                   verbose=False)
+    assert_array_almost_equal(gradex, gradbh, decimal=5)
+
+
+def test_quadtree_similar_point():
+    """Introduce a point into a quad tree where a similar point already exists.
+
+    Test will hang if it doesn't complete.
+    """
+    Xs = []
+
+    # check the case where points are actually different
+    Xs.append(np.array([[1, 2], [3, 4]]))
+    # check the case where points are the same on X axis
+    Xs.append(np.array([[-9.368728, 10.264389], [-9.368728, 11.264389]]))
+    # check the case where points are arbitraryily close on X axis
+    Xs.append(np.array([[-9.368728, 10.264389], [-9.368761, 11.264389]]))
+    # check the case where points are the same on Y axis
+    Xs.append(np.array([[-10.368728, 3.264389], [-11.368761, 3.264389]]))
+    # check the case where points are arbitrarily close on Y axis
+    Xs.append(np.array([[-10.368728, 3.264339], [-11.368761, 3.264389]]))
+    # check the case where points are arbitraryily close on both axes
+    Xs.append(np.array([[-9.368761, 3.264389], [-9.368761, 3.264389]]))
+
+    # check the case where points are arbitraryily close on both axes
+    # close to machine epsilon - x axis
+    Xs.append(np.array([[1, 0.0003817754041], [2, 0.0003817753750]]))
```

Comment:
I think the `dtype` of this arrays should be fixed to `dtype=np.float64` so that the test can be expected to behave consistently across platforms (operating systems and Python architectures).